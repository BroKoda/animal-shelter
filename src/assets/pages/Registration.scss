@import '../variables-mixins';

.registration-container {
  height: 100vh;
  position: relative;

  .registration-box {
    background-color: $gray;
    border-radius: 20px;
    height: 100%;

    .call-to-action-button {
      width: 100%;
    }

    .secondary-button {
      width: 100%;
    }

    .redirect-to-login {
      font-size: $smaller-font-size;
    }
  }
}