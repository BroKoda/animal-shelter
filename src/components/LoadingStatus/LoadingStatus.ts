export const enum LoadingStatus {
  initial = 'initial',
  complete = 'complete',
  loading = 'loading',
  error = 'error'
}