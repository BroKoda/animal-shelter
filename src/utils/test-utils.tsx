import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import type { PreloadedState } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import { setupStore } from '../app/testStore'
import type { TestAppStore, TestRootReducer } from '../app/testStore'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<TestRootReducer>
  store?: TestAppStore
}

export function renderWithProviders(
  ui: JSX.Element,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
): any {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}