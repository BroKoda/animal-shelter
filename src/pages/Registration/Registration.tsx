import React, { ChangeEvent, useCallback, useEffect } from 'react'
import { Link, useNavigate } from 'react-router-dom'
import '../../assets/pages/Registration.scss'
import { useAppDispatch, useAppSelector } from '../../app/hooks'
import {
  registerUser,
  resetRegistration,
  setConfirmPassword,
  setEmail,
  setFirstName, setLastName,
  setPassword
} from './RegistrationSlice'
import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'

const Registration = (): JSX.Element => {
  const navigate = useNavigate()
  const dispatch = useAppDispatch()
  const status = useAppSelector((state) => state.registration.status)
  const { firstName, lastName, email, password, confirmPassword } = useAppSelector((state) => state.registration.userToCreate)

  useEffect(() => {
    if (status === LoadingStatus.complete) {
      dispatch(resetRegistration())
      navigate('/bejelentkezes')
    }
  }, [dispatch, status])

  const handleRegistration = useCallback(() => {
    const user = {
      firstName: firstName,
      lastName: lastName,
      email: email,
      password: password
    }
    void dispatch(registerUser(user))
  }, [dispatch, firstName, lastName, email, password])

  const handleSetFirstName = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    dispatch(setFirstName(event.target?.value))
  }, [dispatch, firstName])

  const handleSetLastName = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    dispatch(setLastName(event.target?.value))
  }, [dispatch, lastName])

  const handleSetEmail = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    dispatch(setEmail(event.target?.value))
  }, [dispatch, email])

  const handleSetPassword = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    dispatch(setPassword(event.target?.value))
  }, [dispatch, password])

  const handleSetConfirmPassword = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    dispatch(setConfirmPassword(event.target?.value))
  }, [dispatch, confirmPassword])

  return (
    <div className="container registration-container">
      <div className="row justify-content-center h-100">
        <div className="col-12 col-md-8 col-lg-6 col-xl-5 align-self-center">
          <div className="container registration-box py-4 px-4">
            <div className="row">
              <div className="col-12">
                <h2>Regisztráció</h2>
                <form className='my-4'>
                  <div className="form-control-container">
                    <div className="form-icon-container">
                      <i className="fa-solid fa-user"></i>
                    </div>
                    <input
                      type="text"
                      className="form-control"
                      placeholder='Vezetéknév'
                      id="lastName"
                      value={lastName ?? ''}
                      onChange={handleSetLastName}
                    />
                  </div>
                  <div className="form-control-container">
                    <div className="form-icon-container">
                      <i className="fa-solid fa-user"></i>
                    </div>
                    <input
                      type="text"
                      className="form-control"
                      placeholder='Keresztnév'
                      id="firstName"
                      value={firstName ?? ''}
                      onChange={handleSetFirstName}
                    />
                  </div>
                  <div className="form-control-container">
                    <div className="form-icon-container">
                      <i className="fa-solid fa-at"></i>
                    </div>
                    <input
                      type="email"
                      className="form-control"
                      placeholder='Email'
                      id="email"
                      value={email ?? ''}
                      onChange={handleSetEmail}
                    />
                  </div>
                  <div className="form-control-container">
                    <div className="form-icon-container">
                      <i className="fa-solid fa-lock"></i>
                    </div>
                    <input
                      type="password"
                      className="form-control"
                      placeholder='Jelszó'
                      id="password"
                      value={password ?? ''}
                      onChange={handleSetPassword}
                    />
                  </div>
                  <div className="form-control-container">
                    <div className="form-icon-container">
                      <i className="fa-solid fa-lock"></i>
                    </div>
                    <input
                      type="password"
                      className="form-control"
                      placeholder='Jelszó mégegyszer'
                      id="confirm-password"
                      value={confirmPassword ?? ''}
                      onChange={handleSetConfirmPassword}
                    />
                  </div>
                </form>
              </div>
            </div>
            <div className='row'>
              <div className='col-12 mb-3 text-end redirect-to-login'>
                <Link to={'/bejelentkezes'}>
                  <span>Van már felhasználód? Lépj be!</span>
                </Link>
              </div>
            </div>
            <div className='row'>
              <div className='col-6'>
                <Link to={'/'}>
                  <button className='button secondary-button'>
                    Vissza
                  </button>
                </Link>
              </div>
              <div className='col-6'>
                <button className='button call-to-action-button' onClick={handleRegistration}>
                  Regisztráció
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

export default Registration
