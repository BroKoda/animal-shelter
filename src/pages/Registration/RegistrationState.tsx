import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'

export interface RegistrationState {
  userToCreate: UserToCreate
  status: LoadingStatus
}

export interface UserToCreate {
  firstName?: string
  lastName?: string
  role?: string
  email?: string
  phone?: string
  password?: string
  confirmPassword?: string
}