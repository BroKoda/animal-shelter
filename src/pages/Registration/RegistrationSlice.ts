import { RegistrationState, UserToCreate } from './RegistrationState'
import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'
import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { RootState } from '../../app/store'
import { createUserWithEmailAndPassword, getAuth } from 'firebase/auth'

const initialState: RegistrationState = {
  userToCreate: {},
  status: LoadingStatus.initial
}

export const registerUser = createAsyncThunk('registerUser', async ({ email, password }: UserToCreate) => {
  if (email != null && password != null) {
    const response = await createUserWithEmailAndPassword(getAuth(), email, password)
    return response
  }
})

const registrationSlice = createSlice({
  name: 'registration',
  initialState,
  reducers: {
    resetRegistration: () => initialState,
    setEmail: (state: RegistrationState, action: PayloadAction<string>) => {
      state.userToCreate.email = action.payload
    },
    setPassword: (state: RegistrationState, action: PayloadAction<string>) => {
      state.userToCreate.password = action.payload
    },
    setConfirmPassword: (state: RegistrationState, action: PayloadAction<string>) => {
      state.userToCreate.confirmPassword = action.payload
    }
  },
  extraReducers: builder => {
    builder
      .addCase(registerUser.pending, (state) => {
        state.status = LoadingStatus.loading
      })
      .addCase(registerUser.fulfilled, (state) => {
        state.status = LoadingStatus.complete
      })
      .addCase(registerUser.rejected, (state) => {
        state.status = LoadingStatus.error
      })
  }
})

export const { resetRegistration, setEmail, setPassword, setConfirmPassword } = registrationSlice.actions
export const selectRegistration = (state: RootState) => state.registration
export default registrationSlice.reducer