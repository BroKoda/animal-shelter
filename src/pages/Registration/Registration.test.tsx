import { LoadingStatus } from '../../components/LoadingStatus/LoadingStatus'
import { renderWithProviders } from '../../utils/test-utils'
import { BrowserRouter } from 'react-router-dom'
import Registration from './Registration'
import { RegistrationState } from './RegistrationState'
import RegistrationSlice, {
  setConfirmPassword,
  setEmail,
  setFirstName,
  setLastName,
  setPassword,
  setPhone
} from './RegistrationSlice'

describe('Contact Page tests', () => {
  const initialState: RegistrationState = {
    userToCreate: {},
    status: LoadingStatus.initial
  }

  it('Should set the registration first name', () => {
    expect(RegistrationSlice(initialState, setFirstName('Test name'))).toEqual({
      userToCreate: {
        firstName: 'Test name'
      },
      status: LoadingStatus.initial
    })
  })

  it('Should set the registration last name', () => {
    expect(RegistrationSlice(initialState, setLastName('Test name'))).toEqual({
      userToCreate: {
        lastName: 'Test name'
      },
      status: LoadingStatus.initial
    })
  })

  it('Should set the registration email', () => {
    expect(RegistrationSlice(initialState, setEmail('test@test.com'))).toEqual({
      userToCreate: {
        email: 'test@test.com'
      },
      status: LoadingStatus.initial
    })
  })

  it('Should set the registration phone', () => {
    expect(RegistrationSlice(initialState, setPhone('00 000 0000'))).toEqual({
      userToCreate: {
        phone: '00 000 0000'
      },
      status: LoadingStatus.initial
    })
  })

  it('Should set the registration password', () => {
    expect(RegistrationSlice(initialState, setPassword('1234a'))).toEqual({
      userToCreate: {
        password: '1234a'
      },
      status: LoadingStatus.initial
    })
  })

  it('Should set the registration password confirm', () => {
    expect(RegistrationSlice(initialState, setConfirmPassword('1234a'))).toEqual({
      userToCreate: {
        confirmPassword: '1234a'
      },
      status: LoadingStatus.initial
    })
  })

  it('Should render registration page', async () => {
    const { container } = renderWithProviders(
      <BrowserRouter>
        <Registration />
      </BrowserRouter>
    )
    expect(container).toMatchSnapshot()
  })
})