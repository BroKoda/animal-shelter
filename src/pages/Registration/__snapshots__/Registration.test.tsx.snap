// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Page tests Should render registration page 1`] = `
<div>
  <div
    class="container registration-container"
  >
    <div
      class="row justify-content-center h-100"
    >
      <div
        class="col-12 col-md-8 col-lg-6 col-xl-5 align-self-center"
      >
        <div
          class="container registration-box py-4 px-4"
        >
          <div
            class="row"
          >
            <div
              class="col-12"
            >
              <h2>
                Regisztráció
              </h2>
              <form
                class="my-4"
              >
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-user"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Vezetéknév
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="lastName"
                    placeholder="Vezetéknév"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-user"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Keresztnév
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="firstName"
                    placeholder="Keresztnév"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-at"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Email
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="email"
                    placeholder="Email"
                    required=""
                    type="email"
                    value=""
                  />
                </div>
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-phone"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Telefon
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="email"
                    placeholder="Telefonszám"
                    required=""
                    type="email"
                    value=""
                  />
                </div>
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-lock"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Jelszó
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="password"
                    placeholder="Jelszó"
                    required=""
                    type="password"
                    value=""
                  />
                </div>
                <div
                  class="form-control-container"
                >
                  <div
                    class="form-icon-container"
                  >
                    <i
                      class="fa-solid fa-lock"
                    />
                    <div
                      class="tooltip-container"
                    >
                      <p>
                        Jelszó ismét
                        <br />
                         Mező kitöltése kötelező!
                      </p>
                    </div>
                  </div>
                  <input
                    class="form-control"
                    id="confirm-password"
                    placeholder="Jelszó mégegyszer"
                    required=""
                    type="password"
                    value=""
                  />
                </div>
              </form>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-12 mb-3 text-end redirect-to-login"
            >
              <a
                href="/bejelentkezes"
              >
                <span>
                  Van már felhasználód? Lépj be!
                </span>
              </a>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col-6"
            >
              <a
                href="/"
              >
                <button
                  class="button secondary-button"
                >
                  Vissza
                </button>
              </a>
            </div>
            <div
              class="col-6"
            >
              <button
                class="button call-to-action-button"
              >
                Regisztráció
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
